File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/api/ApiManager.java
package com.akexorcist.repositoryarchcomponents.api;

import com.akexorcist.repositoryarchcomponents.api.hero.HeroService;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Akexorcist on 10/4/2017 AD.
 */

public class ApiManager {
    private static HeroService heroService;

    public static HeroService getHeroService() {
        if (heroService == null) {
            OkHttpClient.Builder builder = new OkHttpClient.Builder()
                    .addNetworkInterceptor(getDefaultHttpLoggingInterceptor());
            heroService = new Retrofit.Builder()
                    .baseUrl("http://localhost:7777/")
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(new LiveDataCallAdapterFactory())
                    .client(builder.build())
                    .build()
                    .create(HeroService.class);
        }
        return heroService;
    }

    private static HttpLoggingInterceptor getDefaultHttpLoggingInterceptor() {
        return new HttpLoggingInterceptor(new PrettyHttpLogger()).setLevel(HttpLoggingInterceptor.Level.BODY);
    }
}

File: AwesomeArchComponent/app/src/main/java/com/akexorcist/repositoryarchcomponents/api/ApiManager.kt
package com.akexorcist.repositoryarchcomponents.api

import com.akexorcist.repositoryarchcomponents.api.config.ENDPOINT
import com.akexorcist.repositoryarchcomponents.api.hero.HeroService
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


/**
 * Created by Anirut Teerabut on 12/23/2017.
 */
object ApiManager {
    private var heroService: HeroService? = null

    private val defaultHttpLoggingInterceptor: HttpLoggingInterceptor
        get() = HttpLoggingInterceptor().setLevel(HttpLoggingInterceptor.Level.BODY)

    fun getHeroService(): HeroService? {
        if (heroService == null) {
            val builder = OkHttpClient.Builder()
                    .addNetworkInterceptor(defaultHttpLoggingInterceptor)

            heroService = Retrofit.Builder()
                    .baseUrl(ENDPOINT)
                    .addConverterFactory(GsonConverterFactory.create())
                    .addCallAdapterFactory(LiveDataCallAdapterFactory())
                    .client(builder.build())
                    .build()
                    .create(HeroService::class.java)
        }
        return heroService
    }
}